category_theory/opposites.lean

```diff
@@ -0,0 +1,55 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+import category_theory.functor
+import category_theory.products
+import category_theory.types
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂
+
+def op (C : Type u₁) : Type u₁ := C
+
+notation C `ᵒᵖ` := op C
+
+variables {C : Type u₁} [𝒞 : category.{u₁ v₁} C]
+include 𝒞
+
+instance opposite : category.{u₁ v₁} (Cᵒᵖ) := 
+{ Hom     := λ X Y : C, Y ⟶ X,
+  comp    := λ _ _ _ f g, g ≫ f,
+  id      := λ X, 𝟙 X,
+  id_comp := begin /- `obviously'` says: -/ intros, simp end,
+  comp_id := begin /- `obviously'` says: -/ intros, simp end,
+  assoc   := begin /- `obviously'` says: -/ intros, simp end }
+
+namespace functor
+
+variables {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒟
+
+definition opposite (F : C ↝ D) : (Cᵒᵖ) ↝ (Dᵒᵖ) := 
+{ obj      := λ X, F X,
+  map      := λ X Y f, F.map f,
+  map_id   := begin /- `obviously'` says: -/ intros, erw [map_id], refl, end,
+  map_comp := begin /- `obviously'` says: -/ intros, erw [map_comp], refl end }
+
+@[simp] lemma opposite_obj (F : C ↝ D) (X : C) : (F.opposite) X = F X := rfl
+@[simp] lemma opposite_map (F : C ↝ D) {X Y : C} (f : X ⟶ Y) : (F.opposite).map f = F.map f := rfl
+                   
+end functor
+
+variable (C)
+
+definition hom_pairing : (Cᵒᵖ × C) ↝ (Type v₁) := 
```

Comment:
Hm, to me this is not an obvious name. I guess this is the functor mathematicians call `Hom(-,-)`; can we call it `Hom` or `functor.Hom` or `functor.hom`? Or are you trying to call attention to the fact that this is an uncurried functor?

Reply:
This is a fairly common name amongst mathematicians (ha!): the idea is that it is a pairing, like the pairing between a vector space and its dual. I think the common usage is just because "the hom functor for C" doesn't really specify what you mean, but "the hom pairing for C" is unambiguously a "bilinear" thing, i.e. a functor out of `Cᵒᵖ × C`.

I can change it to just `functor.hom`, but I think this is more descriptive.

Reply:
Another thing: I've been thinking that the field `category.Hom` should be `category.hom`. What do you think?

Reply:
I think that's a good idea, and it will free `Hom` up for other uses (such as here?). I think "hom pairing" sounds more like a way to speak the symbol in a mathematical context without the opportunity for notational or typing disambiguation. In our case, I think this should be called `functor.hom` rather than `functor.hom_pairing`, unless you think that there is another definition that has a better claim to the name `functor.hom`. (I thought about the curried version of this functor, but I guess that's usually called `yoneda` or a variant...)